buffer/src/main/java/io/netty5/buffer/adapt/AdaptivePoolingAllocator.java

```diff
@@ -0,0 +1,541 @@
+    private volatile boolean closed;
+
+    public AdaptivePoolingAllocator(boolean direct) {
+        this(MemoryManager.instance(), direct);
+    }
+
+    public AdaptivePoolingAllocator(MemoryManager manager, boolean direct) {
+        allocationType = direct ? StandardAllocationTypes.OFF_HEAP : StandardAllocationTypes.ON_HEAP;
+        this.manager = manager;
+        centralQueue = requireNonNull(createSharedChunkQueue());
+        allocatorControl = new SimpleAllocatorControl(this);
+        magazineExpandLock = new StampedLock();
+        Magazine[] mags = new Magazine[4];
+        for (int i = 0; i < mags.length; i++) {
+            mags[i] = new Magazine(this);
+        }
+        magazines = mags;
+    }
+
+    /**
+     * Create a thread-safe multi-producer, multi-consumer queue to hold chunks that spill over from the
+     * internal Magazines.
+     * <p>
+     * Each Magazine can only hold two chunks at any one time: the chunk it currently allocates from,
+     * and the next-in-line chunk which will be used for allocation once the current one has been used up.
+     * This queue will be used by magazines to share any excess chunks they allocate, so that they don't need to
+     * allocate new chunks when their current and next-in-line chunks have both been used up.
+     * <p>
+     * The simplest implementation of this method is to return a new {@link java.util.concurrent.ConcurrentLinkedQueue}.
+     * However, the {@code CLQ} is unbounded, and this means there's no limit to how many chunks can be cached in this
+     * queue.
+     * <p>
+     * Each chunk in this queue can be up to {@link #MAX_CHUNK_SIZE} in size, so it is recommended to use a bounded
+     * queue to limit the maximum memory usage.
+     * <p>
+     * The default implementation will create a bounded queue with a capacity of {@link #CENTRAL_QUEUE_CAPACITY}.
+     *
+     * @return A new multi-producer, multi-consumer queue.
+     */
+    @NotNull
+    protected Queue<Buffer> createSharedChunkQueue() {
+        return PlatformDependent.newMpmcQueue(CENTRAL_QUEUE_CAPACITY);
+    }
+
+    @Override
+    public final boolean isPooling() {
+        return true;
+    }
+
+    @Override
+    public final AllocationType getAllocationType() {
+        return allocationType;
+    }
+
+    @Override
+    public Buffer allocate(int size) {
+        if (closed) {
+            throw allocatorClosedException();
+        }
+        InternalBufferUtils.assertValidBufferSize(size);
+        if (size <= MAX_CHUNK_SIZE) {
+            int sizeBucket = AllocationStatistics.sizeBucket(size); // Compute outside of Magazine lock for better ILP.
+            int expansions = 0;
+            Magazine[] mags;
+            do {
+                mags = magazines;
+                int mask = mags.length - 1;
+                int index = (int) (threadId(Thread.currentThread()) & mask);
+                for (int i = 0, m = Integer.numberOfTrailingZeros(~mask); i < m; i++) {
```

Comment:
maybe `m = 1 << Integer.numberOfTrailingZeros(~mask)` to iterate all the magazines?
```suggestion
                for (int i = 0, m = 1 << Integer.numberOfTrailingZeros(~mask); i < m; i++) {
```